packages/edit-site/src/components/block-editor/index.js

```diff
@@ -51,28 +58,252 @@ const LAYOUT = {
-					canvasMode: getCanvasMode(),
-					hasPageContentFocus: _hasPageContentFocus(),
-				};
-			},
-			[ setIsInserterOpened ]
+	const { templateType, hasPageContentFocus } = useSelect( ( select ) => {
+		const { getEditedPostType, hasPageContentFocus: _hasPageContentFocus } =
+			unlock( select( editSiteStore ) );
+
+		return {
+			templateType: getEditedPostType(),
+			hasPageContentFocus: _hasPageContentFocus(),
+		};
+	}, [] );
+
+	const settings = useSiteEditorSettings( templateType );
+
+	const BlockEditorComponent = getBlockEditorComponent( templateType );
+
+	return (
+		<BlockEditorComponent
+			settings={ settings }
+			templateType={ templateType }
+		>
+			{ hasPageContentFocus && <DisableNonPageContentBlocks /> }
+			<TemplatePartConverter />
+			<SidebarInspectorFill>
+				<BlockInspector />
+			</SidebarInspectorFill>
+
+			<SiteEditorCanvas
+				settings={ settings }
+				templateType={ templateType }
+			/>
+
+			<ReusableBlocksMenuItems />
+		</BlockEditorComponent>
+	);
+}
+
+/**
+ * Factory to isolate choosing the appropriate block editor
+ * component to handle a given template type/entity.
+ *
+ * @param {string} templateType the template (entity) type being edited
+ * @return {JSX.Element} the block editor component to use.
+ */
+function getBlockEditorComponent( templateType ) {
+	let Component = null;
+
+	switch ( templateType ) {
+		case 'wp_navigation':
+			Component = NavigationBlockEditor;
+			break;
+		default:
+			Component = DefaultBlockEditor;
+			break;
+	}
+
+	return Component;
+}
+
+/**
+ * Block editor component for editing navigation menus.
+ *
+ * Note: Navigation entities require a wrapping Navigation block to provide
+ * them with some basic layout and styling. Therefore we create a "ghost" block
+ * and provide it will a reference to the navigation entity ID being edited.
+ *
+ * In this scenario it is the **block** that handles syncing the entity content
+ * whereas for other entities this is handled by entity block editor.
+ *
+ * @param {number} navigationMenuId the navigation menu ID
+ * @return {[WPBlock[], Function, Function]} The block array and setters.
+ */
+function NavigationBlockEditor( { children, settings, templateType } ) {
+	const noop = () => {};
+
+	useNavigationFocusMode( {
+		templateType,
+	} );
+
+	const navigationMenuId = useEntityId( 'postType', 'wp_navigation' );
+
+	const blocks = useMemo( () => {
+		return [
+			createBlock( 'core/navigation', {
+				ref: navigationMenuId,
+				// As the parent editor is locked with `templateLock`, the template locking
+				// must be explicitly "unset" on the block itself to allow the user to modify
+				// the block's content.
+				templateLock: false,
+			} ),
+		];
+	}, [ navigationMenuId ] );
+
+	return (
+		<ExperimentalBlockEditorProvider
+			settings={ settings }
+			value={ blocks }
+			onInput={ noop }
+			onChange={ noop }
+			useSubRegistry={ false }
+		>
+			{ children }
+		</ExperimentalBlockEditorProvider>
+	);
```

Comment:
For both component we return a pre-configured BlockEditorProvider. The consumer doesn't need to know about the inner workings.